File: 0139-word-break/0139-word-break.py
class TrieNode:
    def __init__(self, val=None):
        self.val = val
        self.children = dict()
        self.last = False


class Trie:
    def __init__(self):
        self.root = TrieNode()
        self.found = dict()


    def insert(self, word):
        node = self.root
        for letter in word:
            if letter not in node.children:
                new = TrieNode(letter)
                node.children[letter] = new
            node = node.children[letter]
        node.last = True


    def search(self, string):
        node = self.root

        for i in range(len(string)):
            if string[i] not in node.children: return False
            if node.children[string[i]].last:
                left = string[i+1:]
                if left not in self.found:
                    self.found[left] = self.search(left)
                if self.found[left]: return True
            node = node.children[string[i]]
        return node.last


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        trie = Trie()

        for word in wordDict:
            trie.insert(word)

        return trie.search(s)

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Trie trie = new Trie();

        for (String word: wordDict) {
            trie.insert(word);
        }

        return trie.search(s);
    }
}

class TrieNode {
    char val;
    Map<Character, TrieNode> children;
    boolean last;

    public TrieNode() {
        this.last = false;
        this.children = new HashMap<>();
    }

    public TrieNode(char val) {
        this.val = val;
        this.last = false;
        this.children = new HashMap<>();
    }
}

class Trie {
    TrieNode root;
    Map<String, Boolean> found;

    public Trie() {
        this.root = new TrieNode();
        this.found = new HashMap<>();
    }

    public void insert(String word) {
        TrieNode node = this.root;

        for (int i = 0; i < word.length(); i++) {
            if (!node.children.containsKey(word.charAt(i))) {
                TrieNode newNode = new TrieNode(word.charAt(i));
                node.children.put(word.charAt(i), newNode);
            }
            node = node.children.get(word.charAt(i));
        }
        node.last = true;
    }

    public boolean search(String str) {
        TrieNode node = this.root;

        for (int i = 0; i < str.length(); i++) {
            if (!node.children.containsKey(str.charAt(i))) {
                return false;
            }
            if (node.children.get(str.charAt(i)).last) {
                String left = str.substring(i + 1, str.length());
                if (!found.containsKey(left)) {
                    found.put(left, search(left));
                }
                if (found.get(left)) {
                    return true;
                }
            }
            node = node.children.get(str.charAt(i));
        }

        return node.last;
    }
}